Mathlib/Topology/Instances/EReal.lean

```diff
@@ -171,6 +170,154 @@ theorem tendsto_nhds_bot_iff_real {α : Type*} {m : α → EReal} {f : Filter α
+    (h : liminf u f ≠ ⊥ ∨ liminf v f ≠ ⊤): (liminf u f) + (liminf v f) ≤ liminf (u + v) f:= by
+  rcases eq_top_or_lt_top (liminf v f) with (v_top | v_ntop)
+  · rcases h with (u_nbot | v_ntop)
+    · rw [add_comm u v, liminf_add_top_of_ne_bot v_top u_nbot]; exact le_top
+    · exfalso; exact v_ntop v_top
+  rcases eq_bot_or_bot_lt (liminf v f) with (v_bot | v_nbot)
+  · rw [v_bot, add_bot]; exact bot_le
+  have liminf_v_real := coe_toReal (ne_of_lt v_ntop) (ne_of_gt v_nbot)
+  apply (ge_iff_le_forall_real_lt _ _).1
+  intros x hx
+  rcases lt_iff_exists_real_btwn.1 hx with ⟨y, ⟨x_lt_y, y_lt_sum⟩⟩
+  have key₁ : (y - liminf v f) < liminf u f := by
+    apply lt_of_lt_of_eq (sub_lt_sub_of_lt_of_le y_lt_sum (le_of_eq (Eq.refl (liminf v f)))
+      (ne_of_gt v_nbot) (ne_of_lt v_ntop))
+    rw [← liminf_v_real, add_sub_cancel_right]
+  have key₂ : liminf v f + x - y < liminf v f := by
+    rw [← liminf_v_real]; norm_cast; norm_cast at x_lt_y; linarith
+  apply le_of_eq_of_le _ (liminf_add_ge_gt₂ key₁ key₂)
+  rw [← liminf_v_real]
+  norm_cast
+  linarith
+
+theorem limsup_le_iff {α : Type _} {f : Filter α} {u : α → EReal} {b : EReal} :
+    limsup u f ≤ b ↔ ∀ c : ℝ, b < c → ∀ᶠ a : α in f, u a ≤ c := by
+  rw [← le_iff_le_forall_real_gt]
+  refine ⟨?_, ?_⟩ <;> intro h c b_lt_c
+  · rcases exists_between_coe_real b_lt_c with ⟨d, b_lt_d, d_lt_c⟩
+    specialize h d b_lt_d
+    have key := Filter.eventually_lt_of_limsup_lt (lt_of_le_of_lt h d_lt_c)
+    apply Filter.mem_of_superset key
+    rw [Set.setOf_subset_setOf]
+    exact fun a h' ↦ le_of_lt h'
+  · rcases eq_or_neBot f with (rfl | _)
+    · simp only [limsup_bot, bot_le]
+    · specialize h c b_lt_c
+      exact @Filter.limsup_const EReal α _ f _ (c : EReal) ▸ limsup_le_limsup h
+
+theorem limsup_le_const_forall {α : Type _} {f : Filter α} {u : α → EReal} {b : EReal}
+    (h : ∀ a : α, u a ≤ b) : limsup u f ≤ b :=
+  limsup_le_iff.2 fun _ b_lt_c ↦ eventually_of_forall (fun a : α ↦ le_trans (h a) (le_of_lt b_lt_c))
+
+theorem const_le_limsup_forall {α : Type _} {f : Filter α} [NeBot f] {u : α → EReal}
+    {b : EReal} (h : ∀ a : α, b ≤ u a) : b ≤ limsup u f :=
+  @Filter.limsup_const EReal α _ f _ b ▸ limsup_le_limsup (eventually_of_forall h)
+
+theorem liminf_le_const_forall {α : Type _} {f : Filter α} [NeBot f] {u : α → EReal}
+    {b : EReal} (h : ∀ a : α, u a ≤ b) :
+    liminf u f ≤ b :=
+  @Filter.liminf_const EReal α _ f _ b ▸ liminf_le_liminf (eventually_of_forall h)
+
+theorem const_le_liminf_forall {α : Type _} {f : Filter α} {u : α → EReal} {b : EReal}
+    (h : ∀ a : α, b ≤ u a) :
+    b ≤ liminf u f := by
+  rcases eq_or_neBot f with (rfl | _)
+  · simp only [liminf_bot, le_top]
+  · exact @Filter.liminf_const EReal α _ f _ b ▸ liminf_le_liminf (eventually_of_forall h)
+
+theorem limsup_max {α : Type _} {f : Filter α} {u v : α → EReal} :
+    limsup (fun a ↦ max (u a) (v a)) f = max (limsup u f) (limsup v f) := by
```

Comment:
You can dualize again to get a `liminf_min` version

Reply:
Done. Thanks.